telegram-purple.c

```diff
@@ -532,23 +530,33 @@ static gulong chat_conversation_typing_signal = 0;
 static void tgprpl_login (PurpleAccount * acct) {
   info ("tgprpl_login(): Purple is telling the prpl to connect the account");
   
+  struct tgl_state *TLS = tgl_state_alloc ();
   PurpleConnection *gc = purple_account_get_connection (acct);
-  
   gc->flags |= PURPLE_CONNECTION_NO_BGCOLOR;
-  
-  struct tgl_state *TLS = tgl_state_alloc ();
   connection_data *conn = connection_data_init (TLS, gc, acct);
   purple_connection_set_protocol_data (gc, conn);
+  tgl_set_ev_base (TLS, conn);
 
   TLS->base_path = get_config_dir (purple_account_get_username (acct));
-  tgl_set_download_directory (TLS, get_download_dir(TLS));
   debug ("base configuration path: '%s'", TLS->base_path);
-  
+  g_mkdir_with_parents (TLS->base_path, 0700);
+
+  const gchar *base = !g_strcmp0 (purple_core_get_ui(), "BitlBee")
+                    ? g_get_tmp_dir() : TLS->base_path;
+  conn->download_dir = g_build_filename (base, "downloads", NULL);
+  debug ("telegram download directory: '%s'", conn->download_dir);
+  tgl_set_download_directory (TLS, conn->download_dir);
+  g_mkdir_with_parents (conn->download_dir, 0700);
+#ifdef DOWNLOAD_URI_BASE
+  conn->download_uri = g_strdup (DOWNLOAD_URI_BASE);
+#else
+  conn->download_uri = g_strconcat ("file://", conn->download_dir, NULL);
+#endif
```

Comment:
Bonus points if you change this to use `tgl_set_download_directory`. This "only" breaks an internal free-memore-counter, so doing it "wrong" isn't too bad.

Reply:
Added `conn->download_dir` is identical with `TLS->downloads_directory` set by `tgl_set_download_directory()` on L548:

```c
544   const gchar *base = !g_strcmp0 (purple_core_get_ui(), "BitlBee")
545                     ? g_get_tmp_dir() : TLS->base_path;
546   conn->download_dir = g_build_filename (base, "downloads", NULL);
547   debug ("telegram download directory: '%s'", conn->download_dir);
548   tgl_set_download_directory (TLS, conn->download_dir);
549   g_mkdir_with_parents (conn->download_dir, 0700);
550 #ifdef DOWNLOAD_URI_BASE
551   conn->download_uri = g_strdup (DOWNLOAD_URI_BASE);
552 #else
553   conn->download_uri = g_strconcat ("file://", conn->download_dir, NULL);
554 #endif
555   debug ("telegram download uri base: '%s'", conn->download_uri);
```

`tgl_set_download_directory()` duplicates the passed-in `conn->download_dir` string argument and stores the result in `TLS->downloads_directory` variable. The resulting duplicated string is freed by `tgl_free_all()` after freeing the new `conn->download_dir` and `conn->download_uri` strings in `connection_data_free()`:

```diff
diff --git a/tgp-structs.c b/tgp-structs.c
index f80dbf6..089fba6 100644
--- a/tgp-structs.c
+++ b/tgp-structs.c
@@ -127,7 +127,9 @@ void *connection_data_free (connection_data *conn) {
   g_hash_table_destroy (conn->id_to_purple_name);
   g_hash_table_destroy (conn->purple_name_to_id);
   g_hash_table_destroy (conn->channel_members);
-  
+  g_free (conn->download_dir);
+  g_free (conn->download_uri);
+
   tgprpl_xfer_free_all (conn);
   g_free (conn->TLS->base_path);
   tgl_free_all (conn->TLS);
```

I think the memory management of `conn->download_dir`, `conn->download_uri` and `TLS->downloads_directory` is correct.

---

We could remove `conn->download_dir` and only use `TLS->downloads_directory` set by `tgl_set_download_directory()`. However, I find the current approach a bit more cleaner and easier to follow because `conn->download_uri` is still needed for storing the download base URI. Also, I'm not sure if `tgl` makes any guarantees regarding `TLS->downloads_directory` so maintaining an _"unnecessary"_ copy `conn->download_dir` in `telegram-purple` makes sense to me.